dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,397 @@
+  }
+
+  /**
+   * Check with etcd if a lease is already expired.
+   * @param lease
+   * @return lease expired
+   */
+  public boolean isLeaseExpired(Lease lease) {
+    return RetryUtils.retryCallable(
+        String.format("Checking IsLeaseExpired, lease:%s",lease.toString()),
+        () -> {
+          LeaseTimeToLiveResponse leaseResp = mClient.getLeaseClient()
+              .timeToLive(lease.mLeaseId, LeaseOption.DEFAULT)
+              .get();
+          return leaseResp.getTTl() <= 0;
+        }, new ExponentialBackoffRetry(100, 500, RETRY_TIMES));
+  }
+
+  /**
+   * Create a childPath with value to a parentPath.
+   * e.g. create "lower_path" under path /upper_path/ to form a
+   * kv pair of /upper_path/lower_path with a given value.
+   * @param parentPath
+   * @param childPath
+   * @param value
+   */
+  public void addChildren(String parentPath, String childPath, byte[] value) {
+    Preconditions.checkState(!StringUtil.isNullOrEmpty(parentPath));
+    Preconditions.checkState(!StringUtil.isNullOrEmpty(childPath));
+    RetryUtils.retryCallable(
+        String.format("Adding child, parentPath:%s, childPath:%s", parentPath, childPath),
+        () -> {
+          String fullPath = parentPath + childPath;
+          PutResponse putResponse = mClient.getKVClient().put(ByteSequence.from(fullPath, StandardCharsets.UTF_8),
+                  ByteSequence.from(value))
+              .get(sDefaultTimeoutInSec, TimeUnit.SECONDS);
+          return true;
+        },
+        new ExponentialBackoffRetry(RETRY_SLEEP_IN_MS, MAX_RETRY_SLEEP_IN_MS, 0));
+  }
+
+  /**
+   * Get list of children path kv pairs from a given parentPath
+   * e.g. get [/upper/lower1 - val1, /upper/lower2 - val2]
+   * under parent path /upper/
+   * @param parentPath parentPath ends with /
+   * @return
+   */
+  public List<KeyValue> getChildren(String parentPath) {
+    return RetryUtils.retryCallable(String.format("Getting children for path:%s", parentPath), () -> {
+      Preconditions.checkState(!StringUtil.isNullOrEmpty(parentPath));
+      GetResponse getResponse = mClient.getKVClient().get(ByteSequence.from(parentPath, StandardCharsets.UTF_8),
+          GetOption.newBuilder().isPrefix(true).build())
+          .get(sDefaultTimeoutInSec, TimeUnit.SECONDS);
+      return getResponse.getKvs();
+    }, new ExponentialBackoffRetry(RETRY_SLEEP_IN_MS, MAX_RETRY_SLEEP_IN_MS, RETRY_TIMES));
+  }
+
+  private void addListenerInternal(
+      String parentPath, StateListener listener, WatchType watchType) {
+    if (mRegisteredWatchers.containsKey(getRegisterWatcherKey(parentPath, watchType))) {
+      LOG.info("Watcher already there for path:{} for children.", parentPath);
+      return;
+    }
+    WatchOption.Builder watchOptBuilder = WatchOption.newBuilder();
+    switch (watchType) {
+      case CHILDREN:
+        String keyRangeEnd = parentPath.substring(0, parentPath.length() - 1)
+            + (char)(parentPath.charAt(parentPath.length() - 1) + 1);
```

Comment:
could you explain in a comment what this manipulation is doing?

Reply:
done